### PR TITLE
add read/write metrics

### PR DESCRIPTION
**Pull request by kvahed**
### Scope & Purpose

Added general read / write io counters to metrics
Added histograms for runtimes in RocksDB for read, insert, replace, update, remove
Added queue time and full URL to HttpCommTask end trace output

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: arangod/GeneralServer/VstCommTask.cpp [hunk 1]
```diff
@@ -374,13 +380,14 @@ void VstCommTask<T>::sendResponse(std::unique_ptr<GeneralResponse> baseRes,
   }
 
   double const totalTime = stat.ELAPSED_SINCE_READ_START();
+  double const queueTime = stat.ELAPSED_WHILE_QUEUED();
```
Comment (jsteemann) [inline arangod/GeneralServer/VstCommTask.cpp]: These lines seem unnecessary now.
```suggestion
```

File: arangod/Scheduler/SupervisedScheduler.cpp [hunk 2]
```diff
@@ -592,7 +592,7 @@ bool SupervisedScheduler::canPullFromQueue(uint64_t queueIndex) const {
   // Otherwise we might have the unlucky case that we first check dequeued,
   // then a job gets done fast (eg dequeued++, done++)
   // and then we read done.
-  uint64_t jobsDone = _jobsDone.load(std::memory_order_acquire);
+   uint64_t jobsDone = _jobsDone.load(std::memory_order_acquire);
```
Comment (jsteemann) [inline arangod/Scheduler/SupervisedScheduler.cpp:595]:
```suggestion
  uint64_t jobsDone = _jobsDone.load(std::memory_order_acquire);
```

File: arangod/GeneralServer/VstCommTask.cpp (hunk 1 above)
Comment (kvahed) [reply arangod/GeneralServer/VstCommTask.cpp]: Outdated. Sorry. Oversight

**Review by jsteemann — APPROVED** [review]

File: arangod/RestServer/MetricsFeature.cpp [hunk 3]
```diff
@@ -64,13 +64,27 @@ void MetricsFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
                      new BooleanParameter(&_export),
                      arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden))
                      .setIntroducedIn(30600);
+  options->addOption("--server.export-read-write-metrics",
+                     "turn metrics for doiument read/write metrics on or off",
```
Comment (jsteemann) [inline arangod/RestServer/MetricsFeature.cpp:68]:
```suggestion
                     "turn metrics for document read/write metrics on or off",
```

File: arangod/RestServer/MetricsFeature.cpp [hunk 4]
```diff
@@ -64,13 +64,27 @@ void MetricsFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
...
+                     new BooleanParameter(&_exportReadWriteMetrics),
+                     arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden))
+                     .setIntroducedIn(30706);
```
Comment (jsteemann) [inline arangod/RestServer/MetricsFeature.cpp:71]:
```suggestion
                     .setIntroducedIn(30707);
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 5]
```diff
@@ -1402,6 +1497,8 @@ Result RocksDBCollection::update(transaction::Methods* trx,
                                  ManagedDocumentResult& resultMdr, OperationOptions& options,
                                  ManagedDocumentResult& previousMdr) {
 
+  ::WriteTimeTracker timeTracker(_statistics._rocksdb_insert_sec, _statistics, options);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1500]:
```suggestion
  ::WriteTimeTracker timeTracker(_statistics._rocksdb_update_sec, _statistics, options);
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 6]
```diff
@@ -1522,6 +1619,8 @@ Result RocksDBCollection::replace(transaction::Methods* trx,
                                   ManagedDocumentResult& resultMdr, OperationOptions& options,
                                   ManagedDocumentResult& previousMdr) {
 
+  ::WriteTimeTracker timeTracker(_statistics._rocksdb_insert_sec, _statistics, options);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1622]:
```suggestion
  ::WriteTimeTracker timeTracker(_statistics._rocksdb_replace_sec, _statistics, options);
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 7]
```diff
@@ -1631,7 +1730,11 @@ Result RocksDBCollection::replace(transaction::Methods* trx,
 Result RocksDBCollection::remove(transaction::Methods& trx, velocypack::Slice slice,
                                  ManagedDocumentResult& previousMdr,
                                  OperationOptions& options) {
+
+  ::WriteTimeTracker timeTracker(_statistics._rocksdb_insert_sec, _statistics, options);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1734]:
```suggestion
  ::WriteTimeTracker timeTracker(_statistics._rocksdb_remove_sec, _statistics, options);
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 8]
```diff
@@ -1659,12 +1762,18 @@ Result RocksDBCollection::remove(transaction::Methods& trx, velocypack::Slice sl
 Result RocksDBCollection::remove(transaction::Methods& trx, LocalDocumentId documentId,
                                  ManagedDocumentResult& previousMdr,
                                  OperationOptions& options) {
+  ::WriteTimeTracker timeTracker(_statistics._rocksdb_insert_sec, _statistics, options);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1765]:
```suggestion
  ::WriteTimeTracker timeTracker(_statistics._rocksdb_remove_sec, _statistics, options);
```

**Review by jsteemann — CHANGES_REQUESTED** [review]
Currently the write operations update the wrong metrics.

We should also backport the tests for the new startup option in `tests/js/client/server_parameters`.

**Comment by kvahed** [comment]
Committed all suggestion. Dangerous auto-merge!

**Review by jsteemann — APPROVED** [review]

File: arangod/RestServer/MetricsFeature.cpp [hunk 9]
```diff
@@ -64,13 +64,27 @@ void MetricsFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
                      new BooleanParameter(&_export),
                      arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden))
                      .setIntroducedIn(30600);
+  options->addOption("--server.export-read-write-metrics",
+                     "turn metrics for document read/write metrics on or off",
+                     new BooleanParameter(&_exportReadWriteMetrics),
+                     arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden))
+                     .setIntroducedIn(30707);
```
Comment (jsteemann) [inline arangod/RestServer/MetricsFeature.cpp:71]:
```suggestion
                     .setIntroducedIn(30708);
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13827/

**Comment by jsteemann** [comment]
Tests blue

**Review by neunhoef — APPROVED** [review]
LGTM

**Pull request merged**